### ISSUE TITLE
PyMQI Architectural Direction - Feature List & Discussion

### ISSUE THREAD

**Issue by Skyler-Altol** [dsuch/pymqi#259]
Hey all,

Figured I'd merge a few conversations that are going on into one big one. Currently, the wishlist from what I can see is as follows:

* https://github.com/dsuch/pymqi/issues/173 - OOP API
* https://github.com/dsuch/pymqi/issues/167 - Clean out `__` attributes and replace with `_`
* Drop Python 2.x support, and replace type hints with Python 3.6+ ones (partially related to https://github.com/dsuch/pymqi/issues/165)
* https://github.com/dsuch/pymqi/issues/166 - Rebuild pymqe in Cython and drop compat code for legacy platforms.
* Improve documentation - looking into the use of something like Sphinx & autodoc would greatly benefit this package. I can personally speak to the frustration of developers at my organisation regarding the lack of an API reference (and without these devs being exposed to the lower level C API, they found it initially very difficult to do stuff that goes beyond the examples provided). They did manage to do what was needed after digging through the comments & docstrings in the PyMQI code however, indicating that the information necessary is already mostly there, we just need to expose it.

If anyone has any objections or suggestions for additions to the above list, please feel free to chime in. I also have one other suggestion that some may not like, which is when IBM drops support for a given version of MQ, we also follow suit (unless a corporate support contract has been provided by Zato, in which case they would need to fork the last relevant version of this package for that version of MQ and then provide hotfixes as necessary). This eliminates a further chunk of legacy compat that often rapidly turns into technical debt. I'm aware that this library has worked with versions of MQ likely older than I am for as many years, but at some point the call has to be made (similiar to Python 2's final end of life call), and things must move forward.

On a side note, IBM has recently announced they're dropping support for V9.0 LTS on September 30th 2021. 9.1 LTS and 9.2 LTS are still in support for the foreseeable future. 

Update:
I'm going to build this major against V9 as a starting point. Any feature that is available to V9 will be assumed in this version. That means all TLS / Message Properties / MQAI / etc will have all their `#ifdef` statements removed. Any features introduced in 9.1 or 9.2 will have ifdefs added to ensure compat between the supported versions.

**Comment by Skyler-Altol** [comment]
In my initial branch I'm going to temporarily rip out the MQAI interface entirely, this can be rebuilt later once the core functionality is operational again under the new major version.

**Comment by Skyler-Altol** [comment]
@SeyfSV @dsuch is there a style guide for this project? PyCharm's default column width is 120, though Python's PEP regarding C styling does explicitly specify that the number of columns should be <80. Looking through pymqe, seems like there's bits where it exceeds 80 but it's always below 120.

What're your thoughts on using 120 as a column limit? Even on small laptop screens, as long as the resolution isn't terrible, you shouldn't have any issues reading the code. With that being said, I'm happy to set it at 80 and let the IDE wrap it for me. Will leave it up to you both.

**Comment by dsuch** [comment]
Thanks for opening the ticket @Skyler-Altol - my comments are below.

* You can find the configuration for flake8 in the project's top-level directory, in the [tox.ini](https://github.com/dsuch/pymqi/blob/master/tox.ini) file. It is a direct copy of what we have in Zato.
* The tox.ini file has a max. line length limit of 160 characters but this is only because sometimes comments and code just cannot be broken into multiple lines and, like you said, using 120 characters per line is just fine.
* I do not have other style guide other than to follow what Zato has. The two links here - [link 1](https://github.com/zatosource/zato/blob/main/code/zato-server/src/zato/server/pubsub/__init__.py), [link 2](https://github.com/zatosource/zato/blob/main/code/zato-cy/src/zato/cy/simpleio.py) are what I consider good examples of clear code that mostly follows PEP8 although it is relaxed per the tox.ini configuration above - please simply make your code look like these.
* The second file is of particular interest because it shows how to use Cython in its so-called pure-Python mode. Essentially, you write Python but it is still valid Cython and it is compiled to C as if it was written in a .pyx file. The benefit is that you can use your IDE's code-completion, for instance.
* I am happy with the idea of dropping support for older MQ versions when it is needed but I do not feel that there is a need to make commitments that we will always do it as soon as IBM does it. There is no need to, simply. Let just life decide - e.g. if we can support MQ 6 without any effort then we can just let it be. But if we do need something new in a few years and we see that MQ 6 gets in our way considerably, then we can drop it.
* I have been using Sphinx for around 7 years and it is good but in Zato we happen to be migrating to [MkDocs](https://squidfunk.github.io/mkdocs-material/) anyway - visually, it surpasses Sphinx, particularly on phones and similar tools, while feature-wise, it is the same. In the longer run, I would prefer to use MkDocs in PyMQI too.
* In terms of auto-generation, an enticing idea may be to generate Python code (dataclasses representing structures like MQCD, MQGMO etc.) directly from C headers - this is doable but, apparently, it cannot be fully automated. The reason is that, again, apparently, there are sometimes situations when artificial padding bytes need to be added for MQ to accept this or that input. Absent of access to MQ source code and without any documentation for its protocol, I see not way around it. This means that if you decide to generate things from C headers, make sure to make room for this kind of customisation.
* Generally, I would not have problems with dropping support for anything else than Linux, Windows and Mac, unless  companies grant us access to their AIX, z/OS or other systems that are not typically used at home. But again, if using our best judgment we can still make sure PyMQI runs on these systems, then let it run - I am just saying that if people would like to make use of someone else's labour then they also need to put in some effort.
* I quite like the overall programmer's API that PyMQI offers in that it is straightforward enough for programmers both experienced and those who just start with Python (let us also remember that many PyMQI users are admins rather than programmers). What I do not like is the notion of passing parameters by reference - the C or C++ way of doing it - and this is definitely something to be changed per #173.
* It probably makes more sense to first change things on the Python level before touching C and Cython. But in either case, it is good that we have tests by @SeyfSV to help in ensuring that any kind of a migration does not break the core functionality (although tests would also need to be updated for sure).
* Finally, if you would like to schedule a call to talk about it all live, please feel free to contact me on dsuch at zato.io as a call may be a more efficient way to communicate.

**Comment by Skyler-Altol** [comment]
Replies:

* "I have been using Sphinx for around 7 years and it is good but in Zato we happen to be migrating to MkDocs anyway - visually, it surpasses Sphinx, particularly on phones and similar tools, while feature-wise, it is the same. In the longer run, I would prefer to use MkDocs in PyMQI too." - Even better, I happen to be a massive fan of MkDocs. Especially when combined with the MkDocs Material theme found here: https://squidfunk.github.io/mkdocs-material/. More than happy to work with MkDocs. 

* Regarding supporting older MQ versions, what you've said makes sense. If it costs us no additional time to support a given version, then I have no issues with supporting it. Perhaps emitting an optional warning to users about a given version of MQ not being supported by IBM may be pertinent (particularly given that some may not follow the life cycles that closely). But that's purely a suggestion. With that being said, I've used V9 as the base for the current fork I'm working on per comments above. Which means that if you tried to use the current code against V6 for example, it's not going to work because it lacks a number of key features present in V9 that are not present in V6. All the structs and whatnot are being configured to initialise to the latest version available as of V9 by default. IBM's code will simply ignore the fields present in the later versions according to documentation, though if C is checking the size of those structs it could get... "interesting".

* Rebuilding the API around not passing parameters by reference may be... difficult. Given that the current incarnation is a thin wrapper around C by its own definition, we'll need to do some more work to abstract that reference design away. I'll have a think about how we may be able to do this (and would greatly appreciate any advice you may have on the matter). 

* I'm thinking that automating the generation of the data structures is a "nice to have" but isn't really worth the investment for the reasons you've outlined. This library isn't that massive in and of itself, so automating that generation (and subsequently losing all the manual fixes applied in the mean time) would be tedious. 

* Regarding system support, absolutely. My org has a fairly large z/OS footprint, but I have veryyyy limited knowledge of z/OS and how it operates. Committing to learning the guts of system Z would be an undertaking larger than overhauling this library. We generally recommend in-house that midrange queue managers act as "concentrators" for client connections and then have a single connection (via cluster or sender/receivers) to the mainframe queue manager to minimise the overhead of connections to the mainframe itself. 

* In general, I'm also a pretty big fan of the API provided. Only thing we need to do as mentioned earlier is write up a proper reference so that programmers can easily find what they need without having to dig through source code to find it. Providing links to the relevant IBM documentation for underlying structures may also be useful for contextual purposes (predominantly because as we all know, the knowledge center is absolutely massive, and finding the specific thing you need when you don't know where to look can be quite... painful).

**Comment by dsuch** [comment]
What I meant by not passing elements by reference is that it would be possibly better to have some kind of a Request structure that queue.put would accept and a Response object that it would return.

For instance:

```python
request = Request()
request.data = 'My Data'
request.md.ReplyToQ = 'My.Queue'

response = queue.put(request)
data = response.data
cid = response.CorrelationId

# etc.
```

There would be still some reference passing going on if one wanted to reuse their MQMD objects or similar but the kind of an API as above would be more object-oriented.

I would like to keep the same names as MQ structures have, e.g. ReplyToQ rather than reply_to_queue because that makes it easy to tell people simply to look up everything in MQ docs.

I see no problems with supporting both the above and what we have today, depending on some kind of a config knob.

As to the warnings - I quite dislike this feature of Python (or C, Java etc.), they are usually difficult to silence out and they would probably amount to noise in logs from a low-level library that PyMQI is.

I am also reminded of one aspect - if we were to drop support for Python 2 then the PyMQI version should be 3.0 rather than 2.0 to avoid any confusion. Or perhaps just keep naming it after the current date of release, like Ubuntu does it, e.g. PyMQI 21.04 to avoid another type of confusion of "whether PyMQI 3.0 works with Python 3.9" or similar.

As to the links to MQ documentation, the question is how do we ensure that they are up to date?

**Comment by Skyler-Altol** [comment]
Before we start building the new Python components (or anything really), one of the lessons I've learned from past API projects is that it's always a good idea to have a specification written up and agreed on. This makes decisions regarding implementation details considerably easier because the design details are already fleshed out.

So regarding the OOP API, there are some structs that make sense to be linked together.

Connection object: CSP, SCO, CD, CNO, (any more?)
Queue object: OD, GMO, PMO + Open/Close options? (Would advise making it a parameter to the constructor of the Queue object that specifies whether you're opening for PUT or GET in plain english.
Pub/Sub objects: Not sure about these ones, this is a weakpoint for me as specified earlier. Can anyone else provide the relevant structs for these? OD would be one, but we could abstract the queue related stuff out.
Message object: MD + misc fields? To be confirmed.

**Comment by dsuch** [comment]
Hello @Skyler-Altol,

I am writing to let you know that, unless there is a sponsor for such works, I currently do not have time to work on this and related tickets.

Thanks.

**Comment by Skyler-Altol** [comment]
Then it seems a quote that was given to me recently is playing itself out. "If it is to be, it's up to me".

I will build the foundation of this myself, and once I feel it's in a reasonable state, I'll request a review. Would this be reasonable and possible @dsuch?

**Comment by dsuch** [comment]
On one hand, I would like to understand what you want to do because I appreciate your taking the time to get involved in everything. On the other hand, I just do not have any way to get involved in it because there are no projects that currently need any changes to PyMQI. Hence, I really recommend a call as otherwise I am afraid I will not have time that you certainly deserve.

**Comment by Skyler-Altol** [comment]
Okay, after talking with some more experienced colleagues I've come to a few realisations. My uh "plan" to overhaul the entire lib from the ground up is likely a bit too ambitious, the primary reason being that this lib is used in a ton of places. I'm bound to make errors during such a redevelopment, and finding all those little bugs would take a while, and a lot of effort from a lot of people. Their suggestion was actually to go with something that dsuch mentioned in an earlier ticket, more specifically building an OOP wrapper around the library rather than rebuilding the library itself. However, before doing that, there are a few things I'd like to fix. 

I've been spending a decent chunk of time recently trying to build something that can parse the C header files and output reasonable Python. So far I've had a bit of a struggle with it, but I'll get there eventually (creating parsers never really was my specialty).

**Comment by dsuch** [comment]
You are right, there is no need to overhaul the whole library and this is why I suggested the call earlier - to be able to understand what you would like to do and what of it can be made part of PyMQI, i.e. a whole rewrite would not be part of it.

If you are looking to learn more about parsing then this is certainly a broad area with dozens of high-quality parser libraries available on PyPI.

In particular, [cffi](https://pypi.org/project/cffi/) is a production-ready library for interfacing with C and they have parsing worked out already. In fact, checking how to wrap MQ APIs using cffi could be a nice experimentation subject.

Though, naturally, please feel free to write your own parser, except that, given that this is a well researched area with many good parsers in existence, currently, I do not see a need for having one in PyMQI.

**Comment by Skyler-Altol** [comment]
Alright after spending a not insignificant amount of time (roughly 7 hours), I have built (mostly by hand) an exact copy of IBM's CMQC from the V9.2 LTS client into a python version. Thanks to doing it by hand, comments were preserved as well. The other headers I am less concerned about as they can be mostly handled by h2py, but CMQC is the big one where there's a ton of ifdefs and other stuff everywhere. Preserving the comments also makes putting new constants in muuuuuuch easier in the future. Using the 9.2 LTS client as a base, one can do a diff on the headers from the CD releases to see what new constants were added, and then add those in quickly by hand.

I really did want to automate this, and I explored a ton of different options. But I just couldn't find something that ticked all the boxes I had. So I made it happen. 

My issue now is that I can see some constants that are defined in our current CMQC that are mysteriously not in the IBM CMQC. Stuff like MQRR_DEFAULT. I'm going to submit a pull request for this new constant file (as it contains all the missing CMQC constants from versions of MQ > V8).

**Comment by dsuch** [comment]
Can you please send a link to the code for now rather than a PR?

**Comment by Skyler-Altol** [comment]
Absolutely, one moment. Will edit this post with a link.

**Comment by Skyler-Altol** [comment]
Just found a typo dammit lol. Fixing.

**Comment by Skyler-Altol** [comment]
A new CMQCFC has now been generated with IBM comments retained as well:

https://gist.github.com/Skyler-Altol/c8de4842ea85f994974354c2ff786544

This was significantly easier due to the significantly smaller number of array conversions required (only 1).

**Comment by Skyler-Altol** [comment]
Just following up guys, @dsuch you've asked people to add new headers in the past in PRs. New functionality like ApplicationName from the MQCNO object (see https://www.ibm.com/docs/en/ibm-mq/9.2?topic=mqi-mqcno-connect-options) can't be implemented without these constants being present. Updating the header files was one of the smaller incremental changes I was looking to do that shouldn't break backwards compat. But I also wanted to retain the useful information from IBM that got stripped out by automated systems in the past. If you could spare a few minutes to just have a quick look at the headers and see what you think, I really would appreciate it.

I know and understand you're busy (and believe me, so am I at work atm), but if the OOP wrapper is even to have a hope of being a thing, then pymqi needs to have the current stuff from IBM in it, otherwise there's no point in me even doing this and I should just drop trying to help entirely.

**Comment by dsuch** [comment]
Hello @Skyler-Altol - this is not about work, I have personal commitments that prevent me from taking part in activities like this project.

The code that you generated looks fine but I will have no time to follow up on anything. Please work with @SeyfSV or @JoshMcMQ. Thank you for your understanding.

**Comment by Skyler-Altol** [comment]
@SeyfSV @JoshMcMQ are either of you in a position to review the work I undertake as part of this? If it is necessary to wait a while for people's lives to be more free then that's understandable. A bit saddening I suppose, but completely understandable. 

For now, I'll start pushing stuff up to my fork as I do it in my own time. Maybe somewhere in the future the stuff I end up pushing to my fork may make its way back into this repository.

**Comment by JoshMcMQ** [comment]
I reviewed both and they look reasonable to me,  I didn't check directly against the .h files.   Once you submit the pull request we can also review the diff between current and replacement.   I agree it will be nice to have the ApplicationName  functionality.   This will be key to have pymqi work with the new uniform cluster feature without needing to mess with the alternate means such a as ENV variables or mqclient.ini settings.   Even if not using uniform clusters it would be nice to have pymqi present either the first or last 28 characters (I prefer the first 28) of the .py file being executed as the ApplicationName instead of the current value, ex python3.exe.

Thank you for your work on these.   I also have had little time with work and personal life to spend much time other than answering some general MQ questions submitted as issues here.

**Comment by SeyfSV** [comment]
Hi, @Skyler-Altol , thank you for your work and interest to project!
1. About constants
Some time ago I wanted to do this, tried h2py and did not find short way to achieve this. You did it - it's cool.
Also I was thinking about exporting constants through pymqe - not good way too. 
If your code/constants passing all our tests it could be merged to master without problem.
But :), some thoughts:
It will be good to have script to automate conversion or just  to have ability to repeat this procedure for new versions.
Also, I have some idea about adding this script to installation process. It will remove need to update pymqi for each new MQ version in repository and fix issue with some OS related variables.

2. About other big changes and enhancements:
As I see, there are no many people who will appreciate you proposition. IBM MQ is not so popular as Kafka and RabbitMQ. I worked more that 10 years with MQ (I love it because I had no big problems with it all this years, it's just works) and now I do not see it in new projects :(. New IT world - new [leaders](https://trends.google.com/trends/explore?date=all&q=%2Fm%2F0401cy,%2Fm%2F0zmynvd,%2Fm%2F0bhc0tk,%2Fm%2F0264f96)
I started using of pymqi for POC for one project, I had issues - I fixed it, now I help other people in using MQ with python, it is just a fun: I like MQ, I like codding on python (I am not programmer) - why not :)
In conclusion: pymqi is just an open-source tool, you can get it, you can use it, you can fix it, you can love it, you can hate it, but no one responsible for spended time. If you wish something - you do this, and propose it for cummunity.

**Comment by dsuch** [comment]
Hello @Skyler-Altol,

the generated code is good but can you, please, also link to the generator that produced it? I could not find a public repo for it in your GH profile.

Thanks.

**Comment by Skyler-Altol** [comment]
Hi guys,

Like I said, the majority of that work was done by hand. I spoke to a bunch of people, and short of writing a full blown grammar specification and associated parser, I could not find anything that was able to parse the C code while also retaining comments. The good news is, as previously stated, as long as the header file version used to build the definitions is maintained, it becomes trivially easy to do a `diff` on the two .h files and compare to see if IBM added any new headers. 

I was genuinely using this:

```
import os
import sys
import clipboard


def main():
    #with open("test.txt", "r") as file:
        #contents = file.readlines()
        
    contents = clipboard.paste().splitlines()
    contents = list(filter(None, contents))
    
    for i in range(len(contents)):
        if "/*" in contents[i]:
            split_line = contents[i].split()
            split_line = split_line[1:-1]
            contents[i] = f"# {' '.join(split_line)}"
        else:
            split_line = contents[i].split()
            contents[i] = f"{split_line[1]} = {split_line[2]}"
    
    thing = ""
    for line in contents:
        thing += f"{line}\r\n"
    
    clipboard.copy(thing)
```

And then going through piece by piece by hand (in CMQC anyway). In CMQCFC there were far less array conversions so it was much easier for this code to process the stuff.

**Comment by dsuch** [comment]
Thanks @Skyler-Altol - I understand it all.

Let me also tell you that I plan to become much more active with PyMQI later this year and then to keep it that way going forward.  I will be sure to return to this discussion.

**Comment by Skyler-Altol** [comment]
Good news @dsuch. I have batphoned an associate who is *far* more knowledgeable in the areas of parsing & language design. He's built a thing that takes in CMQC and spits out a near perfect replica of it in perfect python. It'll still need a bit of tidying up (mainly just deleting some commented stuff for neatness) but my lord it's nuts what this guy has made.

**Comment by dsuch** [comment]
Hi @Skyler-Altol - this is good to know but I am not sure what are you planning to achieve, really?

I asked twice for a call to discuss it and there was no feedback from you. It is difficult for me to treat such an approach seriously.

I see no design, no plan here, and I can understand that you are eager to do something, which is good, but it appears to me that you want to do it on your own, now with another person, without presenting a grander vision except for the desire to have fun while coding, which is good but it is not everything.

It is also difficult for me to get into discussions with "Skyler-Altol". This is serious software for serious purposes and I would appreciate it if I knew who I was talking with, what your name was and what organisation you represented.

Thanks.

**Comment by Skyler-Altol** [comment]
Hi dsuch,

I'll admit I have been on and off. My desire to improve this has been very... up and down lately due to real life dramas. I keep my digital identities and real identities separate, hence my hesitation to participate in calls. I also can't say where I work because I'm genuinely not allowed to officially represent the organisation in this sort of capacity. To do so officially would require a mountain of paperwork and approvals that I likely will never get. But what I can say is this, we're an exceedingly large shop that uses a ton of IBM products. Unfortunately, the vast majority of applications built at my work are written in Java, using products like OpenLiberty or Websphere Application Server as abstraction layers to other middleware products (eg MQ). The result is that pymqi was only really used in one actual business project by the org (and unfortunately, the developers of it borked it quite badly, so that project is being decommissioned and rebuilt using RabbitMQ from what I've heard due to its cost being lower). My interest in this library originally comes from Colin Paice's blog, which showed me the possibilities of what could be done with this. My desire to improve it came from when I was trying to learn how to use it without the requisite C background. 

This other person that got involved isn't involved in the pymqi part of things, I only grabbed him to build the parser (which he treated as an interesting exercise). If nothing else happens, at least a tool will be available for people to keep the headers updated (granted it's written in OCaml so maintaining it may be... difficult. I've raised this and other solutions are being looked at). 

The vision I had realistically is quite simple in theory, but likely won't be anywhere near as easy to implement. A modern python interface that allows for static analysis wherever possible (by using 3.6+ type hints), and an API reference to allow for easier consumption of the library by developers. Most of the stuff at the very top of this thread aligns with exactly that. Rewriting into Cython was rather... questionable due to its backwards compatibility issues, though interesting as a concept. Dropping python 2.x compat simplifies the code base and makes it more maintainable into the future as Python as a language evolves and more new features are added.

Regarding doing it on my own, on the contrary, I really don't want to do it alone because the result would be inferior. I lack the experience with this product as a whole that several others in this community have, yourself included, with various edge cases and other tidbits that have been encountered over the years. 

With all of that being said, I am starting to wonder though... is the community that uses this library big enough to warrant the time investment it would take to do all of this? For you, for me, for everybody who happens to be involved. I'll admit that I'm a perfectionist, and perhaps what already exists is within the realm of "good enough", especially considering I haven't really run into any major issues with this library once I figured out how it works (minus parsing PCF event messages, but that's pretty niche). Perhaps the best thing for me to do is to focus on building the MkDocs stuff you mentioned you preferred at the top.
